### PR TITLE
Implement "until" keyword in gantt charts

### LINKED ISSUE #3173: Add Option `until` / `before` to Gantt Diagram
**Is your feature request related to a problem? Please describe.**
I'd like to set the end of a task by another task / milestone.

**Describe the solution you'd like**
Just like one could set the beginning of a task by `after <taskName>` I want to be able to end it by `before <taskName>` or `until <taskName>`.
It should also support milestones.

**Describe alternatives you've considered**
Specify the date explicitly.

### PR DESCRIPTION
**Pull request by fzag**
## :bookmark_tabs: Summary

Implement keyword `until` for tasks that should run until some specific other task or milestone starts.

Resolves #3173

## :straight_ruler: Design Decisions

The end date of a task could previously be defined either using a duration (e.g. `1d`) or an absolute date (e.g. `2024-01-22`) but not relatively to other tasks/milestones. This is useful in many cases, e.g.: to avoid repeating the end-date of multiple tasks which should end at the same time, rely on critical milestones with specified date, ...

### :clipboard: Tasks

Make sure you

- [x] :book: have read the [contribution guidelines](https://github.com/mermaid-js/mermaid/blob/develop/CONTRIBUTING.md)
- [x] :computer: have added necessary unit/e2e tests.
- [x] :notebook: have added documentation. Make sure [`MERMAID_RELEASE_VERSION`](https://github.com/mermaid-js/mermaid/blob/develop/packages/mermaid/src/docs/community/contributing.md#update-documentation) is used for all new features.
- [x] :bookmark: targeted `develop` branch

### CODE REVIEW

File: packages/mermaid/src/diagrams/gantt/ganttDb.js [hunk 1]
```diff
@@ -343,6 +343,35 @@ const parseDuration = function (str) {
...
   str = str.trim();
 
+  // Test for until
+  const re = /^until\s+([\d\w- ]+)/;
+  const untilStatement = re.exec(str.trim());
+
+  if (untilStatement !== null) {
+    // check all until ids and take the earliest
+    let earliestStartingTask = null;
+    untilStatement[1].split(' ').forEach(function (id) {
```
Comment (sidharthv96) [inline packages/mermaid/src/diagrams/gantt/ganttDb.js:353]: Use named capture groups instead of indexes. 

```js
  const untilStatement = /^until\s+(?<ids>[\d\w- ]+)/.exec(str);
  if (untilStatement) {
    // check all until ids and take the earliest
    let earliestStartingTask = null;
    for (const id of untilStatement.groups.ids.split(' ')) {
```

**Review by sidharthv96 — CHANGES_REQUESTED** [review]
```js

  // Test for until
  const untilStatement = /^until\s+(?<ids>[\d\w- ]+)/.exec(str);
  if (untilStatement) {
    // check all until ids and take the earliest
    let earliestStartingTask = null;
    for (const id of untilStatement.groups.ids.split(' ')) {
      let task = findTaskById(id);
      if (!task) {
        continue;
      }
      if (earliestStartingTask === null || task.startTime < earliestStartingTask.startTime) {
        earliestStartingTask = task;
      }
    }

    if (earliestStartingTask) {
      return earliestStartingTask.startTime;
    }

    // TODO: Not sure if we should use a dynamic date here. 
    const dt = new Date();
    dt.setHours(0, 0, 0, 0);
    return dt;
  }
```

File: packages/mermaid/src/diagrams/gantt/ganttDb.spec.ts [hunk 2]
```diff
@@ -158,6 +158,58 @@ describe('when using the ganttDb', function () {
...
     expect(tasks[2].endTime).toEqual(new Date(2013, 0, 17));
   });
+
+  it('should handle relative end date based on id regardless of sections', function () {
+    ganttDb.setDateFormat('YYYY-MM-DD');
+    ganttDb.addSection('sec1');
+    ganttDb.addTask('task1', 'id1,2013-01-01,until id3');
+    ganttDb.addSection('sec2');
+    ganttDb.addTask('task2', 'id2,2013-01-10,until id3');
+    ganttDb.addTask('task3', 'id3,2013-02-01,2d');
```
[the remark this answers is not in the document]
Comment (fzag) [reply packages/mermaid/src/diagrams/gantt/ganttDb.spec.ts:168]: Thanks for the review and suggestions @sidharthv96, I'll try to find some time later for the e2e and parsing tests. For now, I just addressed the requested changes on the implementation code.

**Comment by sidharthv96** [comment]
image
There is one failing test. Can you please look into that?

**Comment by fzag** [comment]
Just fixed it!
The 2 remaining e2e errors are now due to the e2e test having being modified to test the new feature.

**Comment by sidharthv96** [comment]
Can you please separate the tests to new ones? And add a few more cases as well? Tests have helped catch numerous issues recently, and this feature is significant enough to warrant individual tests that specifically state the purpose of the test. 

Otherwise, in the future, maintainers might not notice subtle failures, or if they do, it'll be difficult to figure out which area is affected.

File: cypress/integration/rendering/gantt.spec.js [hunk 3]
```diff
@@ -14,18 +14,18 @@ describe('Gantt diagram', () => {
...
 
       section A section
-      Completed task            :done,    des1, 2014-01-06,2014-01-08
-      Active task               :active,  des2, 2014-01-09, 3d
-      Future task               :         des3, after des2, 5d
-      Future task2               :         des4, after des3, 5d
+      Completed task                      :done, des1, 2014-01-06,2014-01-08
+      Active task                         :active, des2, 2014-01-09, 3d
+      Future task                         :des3, after des2, 5d
+      Future task2                        :des4, after des3, 5d
```
Comment (sidharthv96) [inline cypress/integration/rendering/gantt.spec.js:20]: These formatting changes should be reverted. These could have been tests cases to check inconsistent spacing. 

So please revert changes to existing test, and add new standalone cases for until.

File: cypress/integration/rendering/gantt.spec.js [hunk 4]
```diff
@@ -14,18 +14,18 @@ describe('Gantt diagram', () => {
...
+      Waiting for task 2                  :after des1, until des4
 
       section Critical tasks
       Completed task in the critical line :crit, done, 2014-01-06,24h
-      Implement parser and jison :crit, done, after des1, 2d
+      Implement parser and jison          :crit, done, after des1, 2d
       Create tests for parser             :crit, active, 3d
       Future task in critical line        :crit, 5d
       Create tests for renderer           :2d
-      Add to mermaid                      :1d
```
Comment (sidharthv96) [inline cypress/integration/rendering/gantt.spec.js:28]: Why is this line removed?

**Review by sidharthv96 — CHANGES_REQUESTED** [review]

File: packages/mermaid/src/diagrams/gantt/ganttDb.js [hunk 5]
```diff
@@ -343,13 +337,36 @@ const parseDuration = function (str) {
...
+      }
+    }
+
+    if (earliestTask) {
+      return earliestTask.startTime;
+    } else {
+      const today = new Date();
+      today.setHours(0, 0, 0, 0);
+      return today;
+    }
```
Comment (sidharthv96) [inline packages/mermaid/src/diagrams/gantt/ganttDb.js:360]: Minor styling suggestion. Here and above for `latestTask`.
```suggestion
    if (earliestTask) {
      return earliestTask.startTime;
    }
    const today = new Date();
    today.setHours(0, 0, 0, 0);
    return today;
```

**Review by sidharthv96 — APPROVED** [review]
Awesome!
